crates/red_knot_python_semantic/src/types/signatures.rs

```diff
@@ -0,0 +1,480 @@
+    fn todo() -> Self {
+        Self {
+            positional_only: Box::default(),
+            positional_or_keyword: Box::default(),
+            variadic: Some(Parameter {
+                name: Name::new_static("args"),
+                annotated_ty: Type::Todo,
+            }),
+            keyword_only: Box::default(),
+            keywords: Some(Parameter {
+                name: Name::new_static("kwargs"),
+                annotated_ty: Type::Todo,
+            }),
+        }
+    }
+
+    fn from_parameters(
+        db: &'db dyn Db,
+        definition: Definition<'db>,
+        parameters: &'db ast::Parameters,
+    ) -> Self {
+        let positional_only = parameters
+            .posonlyargs
+            .iter()
+            .map(|arg| ParameterWithDefault::from_node(db, definition, arg))
+            .collect();
+        let positional_or_keyword = parameters
+            .args
+            .iter()
+            .map(|arg| ParameterWithDefault::from_node(db, definition, arg))
+            .collect();
+        let variadic = parameters
+            .vararg
+            .as_ref()
+            .map(|arg| Parameter::from_node(db, definition, arg));
+        let keyword_only = parameters
+            .kwonlyargs
+            .iter()
+            .map(|arg| ParameterWithDefault::from_node(db, definition, arg))
+            .collect();
+        let keywords = parameters
+            .kwarg
+            .as_ref()
+            .map(|arg| Parameter::from_node(db, definition, arg));
+        Self {
+            positional_only,
+            positional_or_keyword,
+            variadic,
+            keyword_only,
+            keywords,
+        }
+    }
+}
+
+/// A single parameter of a typed signature, with optional default value.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(super) struct ParameterWithDefault<'db> {
+    parameter: Parameter<'db>,
+
+    /// Type of the default value, if any.
+    default_ty: Option<Type<'db>>,
+}
+
+impl<'db> ParameterWithDefault<'db> {
+    fn from_node(
+        db: &'db dyn Db,
+        definition: Definition<'db>,
+        parameter_with_default: &'db ast::ParameterWithDefault,
+    ) -> Self {
+        Self {
+            default_ty: parameter_with_default
+                .default
+                .as_deref()
+                .map(|default| definition_expression_ty(db, definition, default)),
+            parameter: Parameter::from_node(db, definition, &parameter_with_default.parameter),
+        }
+    }
+}
+
+/// A single parameter of a typed signature.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(super) struct Parameter<'db> {
+    /// Parameter name.
+    ///
+    /// It is possible for signatures to be defined in ways that leave positional-only parameters
+    /// nameless (e.g. via `Callable` annotations). In these cases we will synthesize an
+    /// invalid-identifier name for the parameter (for example, `1`, `2`, `3`...).
+    name: Name,
```

Comment:
I slightly prefer explicitly representing synthesized names, e.g., using an `enum` here to make it more resilient. The assumption here is that a real parameter can never be named `1` but the parser could decide to recover from `def a(1="test")` and parse out `1` as an identifier.

Reply:
I don't have strong feelings here; it could be as simple as making this an `Option` and handling a synthesized name at output/display time. What I don't like is the ergonomic impact that then _every_ parameter has to deal with its name being an `Option`, even though only a few parameters (always positional-only ones) can have an empty name.

In general I don't think this representation can/should represent invalid signatures (will discuss that more in the other thread), though invalid-identifier parameter names is a more manageable case where we probably could go ahead and represent it.

Reply:
I switched to using an Option here.